### PR TITLE
HEP: storage network [CI SKIP]

### PR DESCRIPTION
**Pull request by tjjh89017**
Signed-off-by: Date Huang <date.huang@suse.com>

**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**

assign Longhorn to use the specific network to sync replicas.

**Solution:**

https://github.com/tjjh89017/harvester/blob/storage_network_hep/enhancements/20220822-storage-network/20220822-storage-network.md

**Related Issue:**
https://github.com/harvester/harvester/issues/1055

**Test plan:**

### CODE REVIEW

**Comment by tjjh89017** [comment]
Currently, we still need to have some manual operation to stop pods that are attached with Longhorn volumes.
That part will need some advice from you guys.
Thanks a lot

**Comment by tjjh89017** [comment]
add controller to shutdown pods in rancher monitoring.
PR: https://github.com/harvester/harvester/pull/2706

File: enhancements/20220822-storage-network/20220822-storage-network.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+- VM use eth2 and eth3
+- Storage uses eth4 and eth5
+- Create a new VlanConfig with eth4 and eth5
+- Create Network-Attatchment-Definition with Whereabouts config for Storage only
+
+### User Experience In Detail
+
+#### Manual
+
+1. Users have to shutdown all VMs and stop all Pods attached with Longhorn volumes manually before setting Longhorn Storage Network. (Including Grafana and Prometheus)
```
Comment (w13915984028) [inline enhancements/20220822-storage-network/20220822-storage-network.md]: The `Prometheus` and `Grafana` POD are using PVC to store data, and there will be `alertmanager` and `logging`, which also use PVC.

It is a bit complex to stop those PODs, user needs to `scale down` the related deployment/replicaset, meanwhile, the `rancherd/fleet-agent` watches `managedchart` and may revert any downstream change (the `scale down`). That will cause the PODs starts again earlier than expected.

Careful tryout and test is needed.

Comment (tjjh89017) [reply enhancements/20220822-storage-network/20220822-storage-network.md]: I already tested prometheus and grafana.
but i didn't test other new pods.
If you have new pods with LH volume, please help us to describe how to scale down those pods.

Comment (tjjh89017) [reply enhancements/20220822-storage-network/20220822-storage-network.md]: You could refer to this controller

https://github.com/harvester/harvester/pull/2706/files#diff-87504a1be112ef39a8cdaf8814e6bc2da202ca091364097a26d589d62f8c9248

File: enhancements/20220822-storage-network/20220822-storage-network.md [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+Functions in this section will not be implemented in v1.1.0
+
+- add Webhook to check VM status while changing `storage-network` settings
+- add Webhook to verify NAD status while changing `storage-network` settings
+
+### Test plan
+
+1. Shutdown all VMs
+2. install whereabouts charts https://github.com/harvester/harvester/pull/2716
```
Comment (w13915984028) [inline enhancements/20220822-storage-network/20220822-storage-network.md]: Is this chart installed automatically when installation?

File: enhancements/20220822-storage-network/20220822-storage-network.md [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+VM and Storage Network will use different interfaces.
+
+- VM use eth2 and eth3
+- Storage uses eth4 and eth5
+- Create a new VlanConfig with eth4 and eth5
+- Create Network-Attatchment-Definition with Whereabouts config for Storage only
+
+### User Experience In Detail
+
+#### Manual
```
Comment (w13915984028) [inline enhancements/20220822-storage-network/20220822-storage-network.md]: As per #test plan, is `Manual` still supported or should be removed?

File: enhancements/20220822-storage-network/20220822-storage-network.md [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    "bridge":"BRIDGE_NAME",
+    "vlan":VLAN_ID,
+    "ipam":{
+        "type":"whereabouts",
+        "range":"192.168.0.0/24"
+    }
+}
+```
+
+4. add Settings named "storage-network" with NAD name in Harvester settings
```
Comment (w13915984028) [inline enhancements/20220822-storage-network/20220822-storage-network.md]: Is this NAD allowed to change after setting with `storage network` ? or should user remove binding first, then change NAD.

File: enhancements/20220822-storage-network/20220822-storage-network.md [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+5. wait and check Longhorn settings "storage-network" is synced.
+6. check Longhorn `instance-manager-e-XXXX` has an interface named "lhnet1"
```
Comment (w13915984028) [inline enhancements/20220822-storage-network/20220822-storage-network.md:140]: Say the `Controller` is the recommended solution to set `stoarge network`, then `Test plan` needs to be aligned with `Controller`.
When will the `Controller` shutdown internal PODs, when will start again. By which flag/state, the user can start VM again.

File: enhancements/20220822-storage-network/20220822-storage-network.md [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+1. Users have to shotdown all VMs and stop all VMI before setting `storage-network` in Harvester settings
+2. Users need to assign a bridge from VLAN Config and a VLAN ID for storage network (Need to setup IPAM Whereabouts IP range)
+3. Controller will shutdown management pods attached with Longhorn volumes and check all volumes are detached.
+4. Controller will set `storage-network` config to Longhorn settings `storage-network`.
+5. Controller will restart management pods.
+6. Users need to restart VM manually.
+
+### API changes
+
+- New Settings `storage-network`
```
Comment (w13915984028) [inline enhancements/20220822-storage-network/20220822-storage-network.md:67]: The revert of setting seems not described cleary.
Assue the `storage-network` is ready, user wishes to change it back to default mode, namely, no `stoarge-network`.
For most of `Harvester settgings`, they support `set to default value`.
How can user remove existing `storage-network` setting/set back default.

File: enhancements/20220822-storage-network/20220822-storage-network.md (hunk 4 above)
Comment (tjjh89017) [reply enhancements/20220822-storage-network/20220822-storage-network.md]: you cannot change NAD after setting it in LH.

File: enhancements/20220822-storage-network/20220822-storage-network.md (hunk 6 above)
Comment (tjjh89017) [reply enhancements/20220822-storage-network/20220822-storage-network.md:67]: revert config is the same as setting it.

File: enhancements/20220822-storage-network/20220822-storage-network.md (hunk 4 above)
Comment (w13915984028) [reply enhancements/20220822-storage-network/20220822-storage-network.md]: It should be writen cleary in the HEP, is it not allowed by system, or user should not do that.

File: enhancements/20220822-storage-network/20220822-storage-network.md (hunk 6 above)
Comment (w13915984028) [reply enhancements/20220822-storage-network/20220822-storage-network.md:67]: HEP needs to add story for it, wirte down the detailed steps.

Well, HEP is to define black-box view of input/output, user will have clear understanding of the operation steps and outpus.

File: enhancements/20220822-storage-network/20220822-storage-network.md (hunk 3 above)
Comment (tjjh89017) [reply enhancements/20220822-storage-network/20220822-storage-network.md]: removed

File: enhancements/20220822-storage-network/20220822-storage-network.md (hunk 5 above)
Comment (tjjh89017) [reply enhancements/20220822-storage-network/20220822-storage-network.md:140]: fixed

File: enhancements/20220822-storage-network/20220822-storage-network.md (hunk 6 above)
Comment (tjjh89017) [reply enhancements/20220822-storage-network/20220822-storage-network.md:67]: fix

**Comment by bk201** [comment]
@tjjh89017 Could you paste the document to Grammarly and check its suggestions? Thanks!

**Comment by tjjh89017** [comment]
Ok scanned by grammerly already

File: enhancements/20220822-storage-network/20220822-storage-network.md [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+#### Controller
+
+1. Users have to shutdown all VMs and stop all VMI before setting `storage-network` in Harvester settings
+2. Users need to create a VLAN Config with ClusterNetwork name and ensure the cluster network and VLAN config cover all nodes and network connectivity is expected.
+3. Users have to ensure the Cluster Network is configured and VLAN Config will cover all nodes and ensure the network connectivity is working and expected in all nodes.
+4. Users need to assign ClusterNetwork, VLAN ID, IP Range in storage-network settings
+5. Controller will shut down management pods attached to Longhorn volumes and check all volumes are detached.
+6. Controller will set `storage-network` config to Longhorn settings `storage-network`.
+7. Controller will restart management pods.
+8. Users need to restart VM manually.
```
Comment (bk201) [inline enhancements/20220822-storage-network/20220822-storage-network.md]:
```suggestion
8. Users need to start VM manually.
```

File: enhancements/20220822-storage-network/20220822-storage-network.md [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+### API changes
+
+- New Settings `storage-network`
+
+## Design
+
+### Implementation Overview
+
+- Re-use VLAN Bridge from VLAN enhancement
```
Comment (bk201) [inline enhancements/20220822-storage-network/20220822-storage-network.md]:
```suggestion
- Re-use VLAN Bridge from VLAN enhancement.
```

File: enhancements/20220822-storage-network/20220822-storage-network.md [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+- Introduce IPAM [Whereabouts](https://github.com/k8snetworkplumbingwg/whereabouts) to be a global IPAM for Longhorn Pods.
+- Create a Network Attachment Definition with plugin bridge and Whereabouts IPAM
```
Comment (bk201) [inline enhancements/20220822-storage-network/20220822-storage-network.md]:
```suggestion
- Create a Network Attachment Definition with plugin bridge and Whereabouts IPAM.
```

File: enhancements/20220822-storage-network/20220822-storage-network.md [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+- According to [Longhorn storage network limitation](https://longhorn.io/docs/1.3.1/advanced-resources/deploy/storage-network/), users should not modify NAD after setting in Longhorn settings.
+- Controller will create NAD for storage-network, and remove old NAD.
+- Controller will sync storage-network config to Longhorn storage-network setting
```
Comment (bk201) [inline enhancements/20220822-storage-network/20220822-storage-network.md]:
```suggestion
- Controller will sync storage-network config to Longhorn storage-network setting.
```

File: enhancements/20220822-storage-network/20220822-storage-network.md [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+    }
+}
+```
+
+### Test plan
+
+#### apply network
+
+1. Shutdown all VMs
+2. install whereabouts charts https://github.com/harvester/harvester/pull/2716
```
Comment (bk201) [inline enhancements/20220822-storage-network/20220822-storage-network.md]:
```suggestion
2. Install whereabouts charts https://github.com/harvester/harvester/pull/2716
```
Make the first letter to upper case (and the remaining lines).

File: enhancements/20220822-storage-network/20220822-storage-network.md [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+        "range":"192.168.0.0/24"
+    }
+}
+```
+
+### Test plan
+
+#### apply network
+
+1. Shutdown all VMs
```
Comment (bk201) [inline enhancements/20220822-storage-network/20220822-storage-network.md]:
```suggestion
1. Shutdown all VMs.
```
And other lines. Please check if all lists in the document end with a period. We need to be consistent.

File: enhancements/20220822-storage-network/20220822-storage-network.md [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+2. install whereabouts charts https://github.com/harvester/harvester/pull/2716
+3. setup VLAN config and ClusterNetwork
+4. add Settings named "storage-network" with the VLAN ID, ClusterNetwork, and IP Range in Harvester settings
+5. wait and check Longhorn settings "storage-network" is synced.
+6. check Longhorn `instance-manager-e-XXXX` has an interface named "lhnet1"
+7. check settings named "storage-network" status is completed.
+
+#### restore to default
```
Comment (bk201) [inline enhancements/20220822-storage-network/20220822-storage-network.md]:
```suggestion
#### Restore to default
```

File: enhancements/20220822-storage-network/20220822-storage-network.md [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+
+Users want to rollback the config back to default
+
+### User Experience In Detail
+
+#### Controller
+
+1. Users have to shutdown all VMs and stop all VMI before setting `storage-network` in Harvester settings
+2. Users need to create a VLAN Config with ClusterNetwork name and ensure the cluster network and VLAN config cover all nodes and network connectivity is expected.
+3. Users have to ensure the Cluster Network is configured and VLAN Config will cover all nodes and ensure the network connectivity is working and expected in all nodes.
```
Comment (yaocw2020) [inline enhancements/20220822-storage-network/20220822-storage-network.md]: Points 2 and 3 seem duplicated.

File: enhancements/20220822-storage-network/20220822-storage-network.md [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+    - name: rancher-monitoring-alertmanager
+- Harvester VM Import Controller
+    - kind: deployments
+    - namespace: harvester-system
+    - name: harvester-vm-import-controller
+
+#### Settings config example
+```json
+{
+    "vlan": 100,
```
Comment (yaocw2020) [inline enhancements/20220822-storage-network/20220822-storage-network.md:110]: Could the VLAN ID be 0?

Comment (tjjh89017) [reply enhancements/20220822-storage-network/20220822-storage-network.md:110]: current webhook will deny vlan=0.
VLAN should be 1 if users need untagged vlan

File: enhancements/20220822-storage-network/20220822-storage-network.md (hunk 14 above)
Comment (tjjh89017) [reply enhancements/20220822-storage-network/20220822-storage-network.md]: I will remove the duplicate part.

**Review by guangbochen — APPROVED** [review]
LGTM, thanks.

**Review by bk201 — APPROVED** [review]
LGTM

**Pull request merged**